### ISSUE TITLE
How Is Padding Calculated for Conv1D,2D,....

### ISSUE THREAD

**Issue by EnderWiggin14** [tensorflow/tensorflow#47882]
Thank you for submitting a TensorFlow documentation issue. Per our GitHub
policy, we only address code/doc bugs, performance issues, feature requests, and
build/installation issues on GitHub.

The TensorFlow docs are open source! To get involved, read the documentation
contributor guide: https://www.tensorflow.org/community/contribute/docs

## URL(s) with the issue:
[https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv1D]()
[https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/layers/convolutional.py#L375-L512]()
[https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/layers/convolutional.py#L52]()

Please provide a link to the documentation entry, for example:
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/MyMethod

## Description of issue (what needs changing):
Add clear description of how the amount of convolutional padding is computed as well as conv transpose padding.

### Clear description
I'm looking for the padding algorithm used to calculate the amount of padding for "SAME," "VALID," etc, but I cannot find it. I've looked at the above links along with some other links to find the algorithm used, but I've only seen class attributes being assigned strings of what "mode" to use. I could not find anything that looked like an associated function call or a dict of functions which makes it difficult to read the algorithm or even view pre-made documentation. I have looked elsewhere on the web to find methods of computing the padding size. Those methods do work for some if not most cases, but they aren't universal as many of them fail for an input dimension of 9 (9x1 1D-Conv) with a filter size of 3. In this special case, the amount of padding necessary is 3 (the same size as the filters), and the methods I have found have maxed out at a padding of 1 for both sides of the input which will fail. Also, most posts on StackOverflow are in regards to computing the output dimensions of convolutional layers which isn't what I'm looking for. I have also looked at tf.pad(), but there is no translation from "SAME" to an accepted function argument.

This is necessary to be able to port the Tensorflow model to a system where Tensorflow cannot run.

For example, why should someone use this method? How is it useful?
This is necessary to be able to port the Tensorflow model to a system where Tensorflow cannot run. It would also be necessary for any porting to other Deep Learning packages.

### Correct links

Is the link to the source code correct?
As far as I know they are.

### Parameters defined

N/A

### Returns defined

N/A

### Raises listed and defined

N/A

https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/feature_column/categorical_column_with_vocabulary_file#raises

### Usage example

N/A

See the API guide: https://www.tensorflow.org/community/contribute/docs_ref
on how to write testable usage examples.

### Request visuals, if applicable

N/A

### Submit a pull request?

Are you planning to also submit a pull request to fix the issue? See the docs
contributor guide: https://www.tensorflow.org/community/contribute/docs,
docs API guide: https://www.tensorflow.org/community/contribute/docs_ref and the
docs style guide: https://www.tensorflow.org/community/contribute/docs_style

**Comment by Saduf2019** [comment]
@EnderWiggin14 
Kindly open a [stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow) issue for this as it is not a bug or feature request, Please post this kind of support questions at Stackoverflow. There is a big community to support and learn from your questions.
Thanks!

**Comment by EnderWiggin14** [comment]
I will not open a stackoverflow issue because of the times I have, I have not received a single response. This submitted as a documentation issue which it very much so is. It should not be that difficult for someone familiar with the source code to link me to correct page containing the information that I seek.

**Comment by tomerk** [comment]
We'll look at improving the documentation around this.
In the meantime some quick notes:

`padding` gets passed to tf.nn.convolution, and you can see some formulas here (though not sure what formulas you are vs. aren't looking for).:

https://www.tensorflow.org/api_docs/python/tf/nn/convolution

With `valid`, the input is not padded and you just do the convolution.
With explicit paddings (e.g. https://www.tensorflow.org/api_docs/python/tf/nn/conv2d), you specify exactly how much to pad each dimension.

`same` is the more subtle one where just enough padding is added to make the input/output height/width match (e.g. conv2d). I guess there your question is exactly how much padding we add and how it gets split across the before/after?

**Comment by EnderWiggin14** [comment]
@tomerk I am indeed looking for the equations to calculate the total amount of padding used with parameter `same` as well as how Tensorflow splits the front and back padding.

Unfortunately, the link you have provided does not answer my questions; however, I believe I will need that information eventually, so I thank you for it.

**Comment by EnderWiggin14** [comment]
Can I at least get a link to the source code where the amount of padding is calculated?

**Comment by tomerk** [comment]
The convolution padding code is found in the actual kernel implementations (not in the python).

A quick skim of various C++ kernel implementations suggests many of them may be going through here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions-inl.h#L1661

Though it'll probably take building up some familiarity with eigen to fully grok what's going on in this method. (I'm not sure if there's anywhere where the logic is shown or applied in a clearer/simple format)

**Comment by EnderWiggin14** [comment]
Thank you! However, now I'm just confused. Starting at the same link as what you posted

[https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions-inl.h#L1661]()

it appears that the variable names of the switch statement are backwards. It doesn't make sense that "VALID" padding would have any kind of padding at all. My understanding is that, for "VALID" padding, the input dimension must be mappable from integer space to integer space via division by an integer.

Additionally, it doesn't appear that this tells me how much padding is necessary to for "SAME" padding. I'm not looking for the `out_height` or the `out_width`. I'm looking for the padding size calculation and how it is split. My understanding from further searches tells me that Tensorflow splits odd numbered padding with the larger split on the right or bottom.
```
switch (padding_type) {
    case PADDING_VALID: {
      const TensorIndex InputRowsEff = InputRows + padding_top + padding_bottom;
      const TensorIndex InputColsEff = InputCols + padding_left + padding_right;
      out_height = divup(InputRowsEff - kernelRowsEff + 1, row_stride);
      out_width = divup(InputColsEff - kernelColsEff + 1, col_stride);
      break;
    }
    case PADDING_SAME: {
      eigen_assert(!padding_explicit);
      out_height = divup(InputRows, row_stride);
      out_width = divup(InputCols, col_stride);
      break;
    }
```

**Comment by tomerk** [comment]
I believe later parts of that method use the  `out_height` and `out_width` to compute which transformations/contractions to use with eigen. It's not super straightforward to understand which is why I gave that caveat.

+Reedwm who has looked a lot at the details of different convolution paddings and might be able to provide you w/ the info you're looking for more clearly

**Comment by reedwm** [comment]
There used to be documentation on SAME padding, which is accessible with the wayback machine [here](https://web.archive.org/web/20181116104639/https://www.tensorflow.org/api_guides/python/nn) in the section "Convolution". The file with this documentation was deleted in https://github.com/tensorflow/docs/commit/1c81393168e11b106a0b4b7337cee97af3b741b8 (file `nn.md`). We should bring back this padding documentation, updating it to take into account TF2, dilations, and explicit padding.

/CC @lamberta, do you know where this documentation should be placed? Perhaps on [this page](https://www.tensorflow.org/api_docs/python/tf/nn), but I don't know the corresponding source file corresponding to that page. I'm willing to add and update the documentation if you don't know anyone more suitable to update it.

Also, the padding calculations are unfortunately repeated in several places in the code. The easiest-to-understand place is here:

https://github.com/tensorflow/tensorflow/blob/dec8e0b11f4f87693b67e125e67dfbc68d26c205/tensorflow/core/framework/kernel_shape_util.cc#L45-L54

**Comment by lamberta** [comment]
Yes, the module overview page is a good place: https://www.tensorflow.org/api_docs/python/tf/nn
Always an adventure finding the docstring but I believe this is it: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_ops.py#L15

@yashk2810 Is it feasible to add a "View in GitHub" button to the module overview pages?

**Comment by EnderWiggin14** [comment]
@reedwm @lamberta

I believe these sets of links are what I was looking for. I will have to see if the code linked by reedwm will need a ceil function as mentioned in the docstring laberta linked.

Thank you!
